File: typescript/src/examples/08_top_funding.ts
/**
 * 08_top_funding: TOP 10 POSITIVE and BOTTOM 10 NEGATIVE funding rows by
 * funding rate, with annualized percentages.
 */

import { FundingClient } from "../client.js";
import { cliEntry } from "../runner.js";
import type { FundingEntry } from "../types.js";

function annualizedPct(row: FundingEntry): number {
  return row.fundingRate * (24 / row.fundingIntervalHours) * 365 * 100;
}

function signed(value: number, decimals: number): string {
  return (value >= 0 ? "+" : "") + value.toFixed(decimals);
}

function formatRow(row: FundingEntry): string {
  const ann = annualizedPct(row);
  return (
    `${row.exchange.padEnd(12)}  ${row.base.padEnd(10)}  ` +
    `${signed(row.fundingRate, 6).padStart(11)}  ${signed(ann, 4).padStart(8)}%`
  );
}

async function main(): Promise<void> {
  const client = new FundingClient();
  const snap = await client.getSnapshot();

  const positives = snap.data
    .filter((r) => r.fundingRate > 0)
    .sort(
      (a, b) =>
        b.fundingRate - a.fundingRate ||
        a.exchange.localeCompare(b.exchange) ||
        a.base.localeCompare(b.base),
    )
    .slice(0, 10);

  const negatives = snap.data
    .filter((r) => r.fundingRate < 0)
    .sort(
      (a, b) =>
        a.fundingRate - b.fundingRate ||
        a.exchange.localeCompare(b.exchange) ||
        a.base.localeCompare(b.base),
    )
    .slice(0, 10);

  const header =
    `${"exchange".padEnd(12)}  ${"base".padEnd(10)}  ` +
    `${"rate".padStart(11)}  ${"ann%".padStart(9)}`;

  process.stdout.write("TOP 10 POSITIVE\n");
  process.stdout.write(`${header}\n`);
  for (const row of positives) process.stdout.write(`${formatRow(row)}\n`);
  process.stdout.write("\n");
  process.stdout.write("BOTTOM 10 NEGATIVE\n");
  process.stdout.write(`${header}\n`);
  for (const row of negatives) process.stdout.write(`${formatRow(row)}\n`);
}

cliEntry(main);

File: python/examples/08_top_funding.py
"""Top 10 positive and bottom 10 negative funding rates across the universe."""

from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from kairo_funding import FundingClient, cli_entry  # noqa: E402


def annualized_pct(rate: float, interval_hours: int) -> float:
    return rate * (24 / interval_hours) * 365 * 100


def main() -> int:
    with FundingClient() as client:
        snapshot = client.get_snapshot()
    rows = FundingClient.parse_compact_rows(snapshot.get("data", []))
    positives = [r for r in rows if r["funding_rate"] > 0]
    negatives = [r for r in rows if r["funding_rate"] < 0]
    positives.sort(key=lambda r: (-r["funding_rate"], r["exchange"], r["base"]))
    negatives.sort(key=lambda r: (r["funding_rate"], r["exchange"], r["base"]))

    header = f"{'exchange':<12}  {'base':<10}  {'rate':>11}  {'ann%':>9}"
    print("TOP 10 POSITIVE")
    print(header)
    for row in positives[:10]:
        ann = annualized_pct(row["funding_rate"], row["funding_interval_hours"])
        print(
            f"{row['exchange']:<12}  {row['base']:<10}  "
            f"{row['funding_rate']:>+11.6f}  {ann:>+8.4f}%"
        )
    print()
    print("BOTTOM 10 NEGATIVE")
    print(header)
    for row in negatives[:10]:
        ann = annualized_pct(row["funding_rate"], row["funding_interval_hours"])
        print(
            f"{row['exchange']:<12}  {row['base']:<10}  "
            f"{row['funding_rate']:>+11.6f}  {ann:>+8.4f}%"
        )
    return 0


if __name__ == "__main__":
    cli_entry(main)
